src/Elmish.WPF/WpfProgram.fs

```diff
@@ -109,23 +115,58 @@ module WpfProgram =
      *)
     let mutable dispatch = Unchecked.defaultof<Dispatch<'msg>>
 
-    let setState model _ =
+    let elmishDispatcher = Threading.Dispatcher.CurrentDispatcher
+    let mutable threader =
+      if element.Dispatcher = elmishDispatcher then
+        SingleThreaded
+      else
+        Threaded_NoUIDispatch
+
+    // Always must be synchronized with elmishDispatcher (or UI thread in single-threaded case)
+    let setUiState model _syncDispatch =
       match viewModel with
       | None ->
-          let uiDispatch msg = element.Dispatcher.Invoke(fun () -> dispatch msg)
+          let dispatchFromViewModel msg =
+            if element.Dispatcher = Threading.Dispatcher.CurrentDispatcher then
+              match threader with
+              | SingleThreaded -> dispatch msg
+              | Threaded_NoUIDispatch ->
+                let uiWaiter = System.Threading.Tasks.TaskCompletionSource<unit -> unit>()
+                threader <- Threaded_PendingUIDispatch uiWaiter
+
+                let synchronizedUiDispatch () =
+                  threader <- Threaded_UIDispatch uiWaiter
+                  dispatch msg
+                  threader <- Threaded_NoUIDispatch
+
+                elmishDispatcher.InvokeAsync(synchronizedUiDispatch) |> ignore
+                uiWaiter.Task.Result()
+              | Threaded_PendingUIDispatch uiWaiter
+              | Threaded_UIDispatch uiWaiter ->
+                uiWaiter.SetException(exn())
```

Comment:
These are invalid states, and should be completely impossible because this can only execute on a single dispatcher and the dispatcher never leaves the threader in either of these states while exiting. Is there another way to represent that?

Reply:
Maybe comment it and leave an message of the state's presumed impossibility in the exception?

Reply:
```suggestion
                uiWaiter.SetException(exn("Error in core Elmish.WPF threading code. Invalid state reached!"))
```